src/jsonrpc-virtual-machine.cpp

```diff
@@ -197,28 +206,37 @@ const std::string &jsonrpc_mg_mgr::get_remote_parent_address(void) const {
 }
 
 void jsonrpc_mg_mgr::snapshot(void) {
-    // Simulate the snapshot operation as in the gRPC server
-    // If we are forked, we kill the parent server and replace its address with the child's,
-    // then we behave as if we were not forked
+    // If we are forked, discard the pending snapshot
     if (is_forked()) {
-        bool result = false;
-        jsonrpc_request(get_mgr(), get_remote_parent_address(), "shutdown", std::tie(), result);
-        std::swap(m_address[0], m_address[1]);
-        m_address.pop_back();
+        commit();
     }
-    // If we are not forked, we fork a new server as the child and get its remote address
+
+    // To create a snapshot, we fork a new server as the child and get its remote address
     std::string child_address;
     jsonrpc_request(get_mgr(), get_remote_address(), "fork", std::tie(), child_address);
     m_address.push_back(std::move(child_address));
 }
 
+void jsonrpc_mg_mgr::commit() {
+    // If we are not forked, there is no pending snapshot to discard, therefore we are already commited
+    if (!is_forked()) {
+        return;
+    }
+
+    // To commit, we kill the parent server and replace its address with the child's
+    bool result = false;
+    jsonrpc_request(get_mgr(), get_remote_parent_address(), "shutdown", std::tie(), result);
+    std::swap(m_address[0], m_address[1]);
```

Comment:
Here you would call rebind, right?

Reply:
Yes, the call to rebind is just in the lines below, maybe you missed because this was done in a different commit.